### PR TITLE
Move to TypeScript

### PR DESCRIPTION
**Pull request by junderw**
First of many to come.

I'm still pretty new to TypeScript, so I don't really know the ins and outs of how a library should be structured and what to export etc.

I have a branch for TypeScript bitcoinjs-lib which is currently done, and I'm going to move through the dependencies one by one and convert them.

Any thoughts?

### CODE REVIEW

**Comment by dcousens** [comment]
`dist`! :-1: 
I don't think this should be done without tooling/steps for deterministic verification...

If that is too much work,  maybe an intermediate step of typescript bindings?

**Comment by junderw** [comment]
dist! :stuck_out_tongue_closed_eyes: 

Hmm, I wonder if committing the js files would work?

**Comment by junderw** [comment]
Just added the type definitions now.

Not sure how you could test the definitions.

**Comment by junderw** [comment]
push failed because it somehow thought the typeScript branch was on bitcoinjs when I'm sending my PR from my personal fork... weird.

So I have rewritten this PR to be more like the PR on the main repo:

Committed TS and JS both to github, be wary of diffs to the JS before publishing.

A review would probably be:
- check the TS from top to bottom
- npm run build
- check if the built JS files caused a diff
- check the built JS files from top to bottom for any weird stuff Microsoft might have injected.

Once we get all this on master, subsequent PRs can be easy to review the TS and JS without having to do a full review each time.

**Comment by junderw** [comment]
Switched from standardJS to the TypeScript standard tslint.

Prettier for formatting, linter for logic problems.

This looks good to merge, but I will wait for a couple more eyes on it before moving forward.

**Comment by junderw** [comment]
@d-yokoi Could you review please?

File: ts-src/bip32.ts [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+const UINT256_TYPE = typeforce.BufferN(32);
+const NETWORK_TYPE = typeforce.compile({
+  wif: typeforce.UInt8,
+  bip32: {
+    public: typeforce.UInt32,
+    private: typeforce.UInt32,
+  },
+});
+
+const BITCOIN = {
```
Comment (d-yokoi) [inline ts-src/bip32.ts]: You might want to declare like this so you don't need to append `as Network` every time.
```
const BITCOIN: Network = {
```

Comment (junderw) [reply ts-src/bip32.ts]: Ok, done. I also changed some as Buffers to use ! instead.

**Comment by d-yokoi** [comment]
My proposal to refactor initialization.
https://github.com/junderw/bip32/pull/1

**Comment by junderw** [comment]
Ok, thanks @d-yokoi for the help.

This looks good to merge.

**Comment by junderw** [comment]
Will publish as 2.0.0 on npm and tag it.

**Pull request merged**